src/idl_gen_julia.cpp

```diff
@@ -0,0 +1,820 @@
+// Documentation comments
+const CommentConfig JuliaCommentConfig = { "#=", "# ", "=#" };
+
+// Class to represent a dependency graph
+// with a topological sort operation
+class DepGraph {
+ public:
+  // adjacency list
+  std::vector<std::vector<unsigned int> *> adj;
+  // node names
+  std::vector<std::string> nodes;
+  unsigned int GetOrCreateNodeID(std::string name) {
+    auto it = std::find(nodes.begin(), nodes.end(), name);
+    if (it == nodes.end()) {
+      nodes.push_back(name);
+      return nodes.size() - 1;
+    }
+    return std::distance(nodes.begin(), it);
+  }
+  DepGraph() {}
+  ~DepGraph() {
+    for (unsigned int i = 0; i < adj.size(); i++) delete adj[i];
+  }
+  void AddDep(std::string parent, std::string child) {
+    unsigned int p = GetOrCreateNodeID(parent);
+    unsigned int c = GetOrCreateNodeID(child);
+    // n = max(p, c)
+    unsigned int n = p > c ? p : c;
+    // if we just created new node ids,
+    // grow the adjacency list to the new size
+    if (n >= adj.size()) {
+      for (unsigned int i = adj.size(); i <= n; i++)
+        adj.push_back(new std::vector<unsigned int>());
+    }
+    adj[p]->push_back(c);
+  }
+  // topological sort
+  bool TopSort(unsigned int v, bool visited[], bool visiting[],
+               std::list<unsigned int> &stack) {
+    bool has_cycle = false;
+    visited[v] = true;
+    visiting[v] = true;
+    for (auto it = adj[v]->begin(); it != adj[v]->end(); ++it) {
+      if (!visited[*it]) has_cycle = TopSort(*it, visited, visiting, stack);
+      if (visiting[*it] && nodes[v] != nodes[*it]) {
+        printf(
+            "warning: %s\n",
+            ("Circular reference detected (" + nodes[v] + " -> " + nodes[*it] +
+             "). " + "Julia does not currently suport such definitions.")
+                .c_str());
+        has_cycle = true;
+      }
+    }
+    visiting[v] = false;
+    stack.push_back(v);
+    return has_cycle;
+  }
+  // Topological sort of dependency graph, so we can include
+  // julia files in the right order
+  std::vector<std::string> TopSort() {
+    std::vector<std::string> sorted_nodes;
+    std::list<unsigned int> stack;
+    unsigned int n = adj.size();
+    bool *visited = new bool[n];
+    bool *visiting = new bool[n];
+    for (unsigned int i = 0; i < n; i++) {
+      visited[i] = false;
+      visiting[i] = false;
+    }
+
+    bool has_cycle = false;
+    for (unsigned int i = 0; i < n; i++)
+      if (!visited[i]) has_cycle |= TopSort(i, visited, visiting, stack);
+
+    while (!stack.empty()) {
+      unsigned int id = stack.back();
+      // don't return nodes on which nothing depends
+      if (!adj[id]->empty()) sorted_nodes.push_back(nodes[id]);
+      stack.pop_back();
+    }
+    return sorted_nodes;
+  }
+};
+
+// Singleton class which keeps track of generated modules
+// and their dependencies. A singleton class is necessary
+// since our generator is run multiple times with
+// multiple different .fbs files, and we want to preserve
```

Comment:
This is a bit unfortunate, since for all languages code generation from a single .fbs is supposed to be a stateless operation. If it is so that we can't get around this, we better document super clearly that code generation for Julia has additional rules.

Reply:
Unfortunately, if FBS files depend on other ones, then yes this is true until Julia has forward declaration of types. https://github.com/JuliaLang/julia/issues/269

Reply:
I'm just looking at updating the documentation for this and it seems a bit unclear. The default seems to be that code will be generated from included files unless the `--no-includes` option is specified. However in `Schemas.md` it says 
```
When using the `flatc` compiler to generate code for schema definitions,
only definitions in the current file will be generated, not those from the
included files (those you still generate separately).
```
What should I write where?

Reply:
100% agree, this is the point of code reviews - making sure new stuff fits as closely as possible :)

Just so I’m clear on the desired behaviour, let’s take `namespace_test` as an example. There are two `.fbs` files - `namespace_test1.fbs` and `namespace_test2.fbs`. `namespace_test1.fbs` defines a namespace `NamespaceA.NamespaceB`. `namespace_test2.fbs` includes `namespace_test1.fbs` and defines some other things in `NamespaceA`. With the singleton class, the Julia code generated looks like this
```julia
module NamespaceA
    include(“NamespaceB/NamespaceB.jl”)
    include(“TableInFirstNS.jl”)
    include(“SecondTableInA.jl”)
end
```
Without the singleton class, the `include(“NamespaceB/NamespaceB.jl”)` is lost and the `TableInFirstNS` type is missing definititions for the types of its fields. 

To be explicit, the definitions for `NamespaceB` are not generated here, but we need to keep track of which files to include, hence the need for the `ModuleTable`. This issue could potentially be resolved by doing away with the object API altogether as discussed elsewhere in this PR.

Reply:
You should be able to call `flatc` on `namespace_test1.fbs`, and get a stand-alone working Julia file (or files). Then you compile `namespace_test2.fbs` which produces include statements to pull in the previous files, and only definitions for what is in `namespace_test2.fbs`.

If you look at existing generators, they check `.generated` on any struct/table/enum to skip them, these are the ones from the includes.

I am not familiar with how Julia does namespaces so I can't tell why it is not possible to follow this pattern. Some language allow namespaces to be "re-opened" (like C++, JS..), others use prefixing (like Python) to achieve the same.

Reply:
I'm with you. I think I may have a solution to this and the inplace reading/writing. I'm on holiday this week but keen to get this sorted early in the new year.

Reply:
Spent a bit of time looking at this today. It doesn't appear that Julia modules can be re-opened in the same way that C++ namespaces can. Also, Julia does not use prefixing - files must be explicitly included. Since there are no forward declarations, the generated files must also be included in the right order. It does seem like there should be a way to do this, but the solution is not obvious to me and I suspect it is quite hairy and involved.

Reply:
Hmm, I'm afraid I am not much help there. Have you asked around with Julia experts how they would solve this?

Reply:
I haven't, but I don't feel well acquainted enough with any of the Julia package experts to bug them. @quinnj might have better contacts than me. 

The only solution I can think of is to save the module hierarchy in a structured format. When you need to add something into the hierarchy, merge it in and then `eval()` everything into existence. Something like this:
```
# NamespaceTest1.jl
module NamespaceTest1

function getmoduletree()
(
       name="NamespaceA",
       includes=[
          (name="NamespaceB",
            includes="NamespaceB/NamespaceB.jl"),
          "TableInFirstNS.jl"
      ]
)
end

end

# namespace_test1_generated.jl

function loadmoduletree(root)
         quote 
              module $(root.name)
                   $([isa(String, n) ? :(include("$n")) : loadmoduletree(n) for n in root.includes]...)
              end
         end
end

include("NamespaceTest1.jl")
import .NamespaceTest1
root = NamespaceTest1.getmoduletree()

eval(loadmoduletree(root))

# namespace_test2_generated.jl

function mergemoduletrees(tree1, tree2)
         # merge two module hierarchies together
end

include("NamespaceTest1.jl")
import .NamespaceTest1
root = NamespaceTest1.getmoduletree()

include("NamespaceTest2.jl")
import .NamespaceTest2
root = mergemoduletrees(root, NamespaceTest2.getmoduletree())

eval(loadmoduletree(root))
```
This is not pretty, and means we are circumventing Julia's module system, which makes it tricky to work with the generated code. That said, if you think the benefits outweigh these costs I'm happy to give this a go.

Reply:
thanks to a slack session with @quinnj yesterday, have come up with a solution to this. Singleton class is gone. Final things left to do are the API for incremental building and adding Docker tests for Julia.